BEIMA.Backend.Test/MongoConnectorTest.cs

```diff
@@ -0,0 +1,227 @@
+        {
+        }
+
+        [SetUp]
+        public void SetUp()
+        {
+        }
+
+        [TearDown]
+        public void TearDown()
+        {
+        }
+
+        [Test]
+        public void InitialState_Action_ExpectedResult()
+        {
+            Assert.IsTrue(true);
+        }
+
+        [Test]
+        public void ConnectorNotCreated_CallConstructorFiveTimes_FiveInstancesAreEqual()
+        {
+            //Tests to make sure singleton is working
+            var connector1 = MongoConnector.Instance;
+            var connector2 = MongoConnector.Instance;
+            var connector3 = MongoConnector.Instance;
+            var connector4 = MongoConnector.Instance;
+            var connector5 = MongoConnector.Instance;
+            Assert.IsTrue(connector1 == connector2);
+            Assert.IsTrue(connector1 == connector3);
+            Assert.IsTrue(connector1 == connector4);
+            Assert.IsTrue(connector1 == connector5);
+        }
+
+        [Test]
+        public void ConnectorNotCreated_CallConstructorOnce_InstanceConnectedToCorrectEnvironment()
+        {
+            var mongo = MongoConnector.Instance;
+            if (Environment.GetEnvironmentVariable("CurrentEnv") == "dev-local")
+            {
+                Assert.IsTrue(mongo.CurrentServerType == ServerType.Local);
+            }
+            else
+            {
+                Assert.IsTrue(mongo.CurrentServerType == ServerType.Cloud);
+            }
+        }
+
+        [Test]
+        public void ConnectorCreated_GetDeviceGivenValidDeviceId_DeviceDocumentReturned()
+        {
+            var mongo = MongoConnector.Instance;
+            //Setup (inserting a document)
+            BsonDocument doc = new BsonDocument {
+                { "deviceTypeId", "a" },
+                { "serialNumber", "a12345"}
+            };
+            var insertResult = mongo.InsertDevice(doc);
+            Assert.IsNotNull(insertResult);
+            Assert.IsTrue(insertResult is ObjectId);
+
+            //Test (retrieve the document)
+            var retrievedDoc = mongo.GetDevice((ObjectId) doc["_id"]);
+            Assert.IsNotNull(retrievedDoc);
+            Assert.IsTrue(retrievedDoc is BsonDocument);
+        }
+
+        [Test]
+        public void ConnectorCreated_GetAllDevices_DeviceDocumentListReturned()
+        {
+            var mongo = MongoConnector.Instance;
+            //Setup (inserting a document)
+            BsonDocument doc = new BsonDocument {
+                { "deviceTypeId", "a" },
+                { "serialNumber", "a12345"}
+            };
+            var insertResult = mongo.InsertDevice(doc);
+            Assert.IsNotNull(insertResult);
+            Assert.IsTrue(insertResult is ObjectId);
+
+            //Test (retrieve all documents)
+            var retrievedDocs = mongo.GetAllDevices();
+            Assert.IsNotNull(retrievedDocs);
+            Assert.IsTrue(retrievedDocs is List<BsonDocument>);
```

Comment:
We should check individual list items as well.

Reply:
The List<BsonDocument> is already type safe for BsonDocuments.

Reply:
I meant like, checking to see that the items in the list have the actual data we expect them to have.

Reply:
I agree, the more testing we can do the better.

Reply:
After a discussion with @AshlynAdamson offline, we have determined that this is better as a functional test. The collection is in an uncertain state when this unit test is run. Because of this uncertainty, it is better if we make it an FT so the state is actually certain.